### DISCUSSION TITLE
Finding simple memory leaks

### DISCUSSION THREAD

**Discussion by romek-codes** [larastan/larastan#1707] [Ideas]
Let's say i have the following code:
```
public function sendMail(SendMailRequest $request): JsonResponse
    {
    // Code
    }
```
And in SendMailRequest Class i've added this:
```
public function rules(SendMailRequest $request): array
    {
    // Validation rules
    }
```

This will of course cause a memory leak when calling the sendMail method.
Would it be possible to find simple memory leaks like this and maybe others when running Larastan?
I think that would be a good feature to have, should probably even show up when running Larastan on level 1.

I'd like to contribute and implement this myself but I wouldn't even know where to start.

**Comment by mad-briller** [comment]
i might be missing something simple here but can you explain to me how the code example you provided causes a memory leak?

**Reply by romek-codes** [reply]
I'm not sure if memory leak is the correct word for this, but in this specific example if you were to send a request to the sendMail endpoint you would get an error like this:
```
Allowed memory size of 33554432 bytes exhausted (tried to allocate 43148176 bytes) in php
``` 
and it would return a 500 status code, because the sendMailRequest class would inject the sendMailRequest class into itself infinitely from my understanding.

**Reply by mad-briller** [reply]
so you want to detect unbounded recursion? that's probably not a trivial thing to detect beyond the simplest cases to be honest 

could you provide a more fleshed out example that leaves less to the imagination? it would help to be able to visualize what could be detected

**Reply by szepeviktor** [reply]
An _infinite loop_? ♾️

**Reply by romek-codes** [reply]
> so you want to detect unbounded recursion? that's probably not a trivial thing to detect beyond the simplest cases to be honest 
> 
> could you provide a more fleshed out example that leaves less to the imagination? it would help to be able to visualize what could be detected

The example i've showcased is actually one that happened to me yesterday while refactoring (brain ran out of memory), if i hadn't called that endpoint i wouldn't have known that i have a fatal bug in my code. There are probably more examples but that's honestly the first time i've stumbled upon such a problem in laravel. I know that in Angular if a [dependency of a service depends on the service itself](https://angular.io/errors/NG0200) it causes a similar issue, and i'm not entirely sure if that also happens in Laravel but catching errors like this would certainly be a very nice thing, but I can't tell how hard that would be to implement.
